packages/health-and-info/src/index.ts

```diff
@@ -0,0 +1,195 @@
+        .get(component.url)
+        .agent(agent)
+        .timeout(timeout)
+        .retry(retries, (err, res) => {
+          attemptsCount += 1
+          if (err) {
+            this.logger.info(
+              `Attempting to get health of external service ${component.name}, got: ${err.code} - ${err.message}`,
+            )
+          }
+        })
+
+      return {
+        name: component.name,
+        status: response.status === 200 ? 'UP' : 'DOWN',
+      }
+    } catch (error: unknown) {
+      this.logger.warn(
+        `Failed getting health of external service ${component.name}`,
+        (error as superagent.ResponseError).stack,
+      )
+
+      return {
+        name: component.name,
+        status: 'DOWN',
+        details: {
+          status: (error as superagent.ResponseError).status,
+          message: (error as superagent.ResponseError).message,
+          attempts: attemptsCount,
+        },
+      }
+    }
+  }
+
+  private checkCustomComponents = (components: Partial<ExternalComponent | CustomComponent>[]) => {
+    return components
+      .filter(component => 'healthCheckFunction' in component)
+      .map(component => ({ ...this.defaultOptions.components, ...component }) as CustomComponent)
+      .filter(component => component.enabled)
+      .map(this.checkCustomComponent)
+  }
+
+  private checkCustomComponent = async (component: CustomComponent): Promise<ComponentCheckResult> => {
+    try {
+      const result = await component.healthCheckFunction()
+
+      if (typeof result === 'boolean') {
+        return {
+          name: component.name,
+          status: result ? 'UP' : 'DOWN',
+        }
+      }
+
+      return {
+        ...result,
+        name: component.name,
+      }
+    } catch (error: unknown) {
+      return {
+        name: component.name,
+        status: 'DOWN',
+        details: {
+          message: (error as Error).message,
+        },
+      }
+    }
+  }
+
+  /** Run and return component health checks */
+  check = async (): Promise<HealthCheckResult> => {
+    const deploymentInfo = this.getDeploymentInfo(this.options.applicationInfo)
+    if (this.options.components?.length) {
+      const componentHealth = await Promise.all([
+        ...this.checkCustomComponents(this.options.components),
+        ...this.checkExternalComponents(this.options.components),
+      ])
+
+      return {
+        status: componentHealth.some(component => component.status === 'DOWN') ? 'DOWN' : 'UP',
+        components: componentHealth,
+        ...deploymentInfo,
+      }
+    }
+
+    return {
+      status: 'UP',
+      ...deploymentInfo,
+    }
+  }
+
+  /**
+   * Middleware handler for the `/health` endpoint.
+   * Generates a status report on individual components and the overall service health.
+   * Responds with HTTP status 200 if all components are healthy, or 500 if any are unhealthy.
+   */
+  private health = async (req: Request, res: Response, next: NextFunction) => {
```

Comment:
**suggestion:** This module is pretty long and a bit mishmash of different responsibilities.. bit of routing, api calling, configuring, a lot of error handling etc.. and its lead to a really large test class.

The original health check code had a nice feature, the actual `/health` endpoint was configured with a list of checks.. very detached from what type of checks were taking place which lead to very easy to test code..

could do something similar to that here by allowing registering of a very abstract type: 
```ts
interface HealthCheckResult {
  name: string
  status: 'UP' | 'DOWN'
  message: unknown
}
export type HealthCheck = () => Promise<HealthCheckResult>
```

You could then create and test a `apiHealthCheck` module separately, which implements `HealthCheck`:
```
const healthCheckOptions = {
  applicationInfo: services.applicationInfo,
  checks: [
    apiHealthCheck({
      name: 'hmpps-auth',
      url: `${config.apis.hmppsAuth.url}/health/ping`,
      enabled: true,
    }),
    apiHealthCheck({
      name: 'token-verification',
      url: `${config.apis.tokenVerification.url}/health/ping`,
      enabled: config.apis.tokenVerification.enabled,
    }),
  ],
}
````

This would mean you could remove all the custom code around trying to handle different types of checks, if someone wanted a custom health check for redis or whatevs.. they would just need to implement a `HealthCheck`

Reply:
I like this suggestion 😄 It does seem neater to me, and opens up the opportunity for adding additional types of checks in future (albeit no idea what these would be currently) without breaking existing implementations/changing up the interfaces heavily.

Reply:
Implemented in all the commits past https://github.com/ministryofjustice/hmpps-typescript-lib/pull/7/commits/c067ed06dfe19251371b33c55a37a9021ce24eec